File: MoonEngine-Root/Base/src/main/java/com/greentree/engine/moon/base/scene/base/SceneManagerBase.java
package com.greentree.engine.moon.base.scene.base;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Objects;
import java.util.ServiceLoader;

import com.greentree.engine.moon.base.scene.EnginePropertiesWorldComponent;
import com.greentree.engine.moon.base.scene.Scene;
import com.greentree.engine.moon.base.scene.SceneManager;
import com.greentree.engine.moon.ecs.World;
import com.greentree.engine.moon.ecs.WorldComponent;
import com.greentree.engine.moon.ecs.system.DestroySystem;
import com.greentree.engine.moon.ecs.system.ECSSystem;
import com.greentree.engine.moon.ecs.system.InitSystem;
import com.greentree.engine.moon.ecs.system.UpdateSystem;
import com.greentree.engine.moon.modules.EngineProperties;

public final class SceneManagerBase implements SceneManager {
	
	private Scene nextScene;
	private boolean loadScene = false;
	private World currentWorld;
	private ECSSystem currentSystems;
	private final Collection<ECSSystem> globalSystems = new ArrayList<>();
	
	{
		final var autoAddGlobalSystems = ServiceLoader.load(ECSSystem.class);
		for(var system : autoAddGlobalSystems)
			globalSystems.add(system);
	}
	
	private final EngineProperties properties;
	
	public SceneManagerBase(EngineProperties properties) {
		this.properties = properties;
	}
	
	@Override
	public void addGlobalSystem(ECSSystem system) {
		System.out.println(system);
		Objects.requireNonNull(system);
		globalSystems.add(system);
	}
	
	@Override
	public synchronized void set(Scene scene) {
		if(loadScene)
			throw new IllegalArgumentException("set scene on load scene");
		nextScene = scene;
	}
	
	public void update() {
		if(nextScene != null)
			loadScene();
		if(currentSystems instanceof UpdateSystem usystem)
			usystem.update();
	}
	
	public void clearScene() {
		if(currentSystems instanceof DestroySystem dsystem)
			dsystem.destroy();
	}
	
	private synchronized void loadScene() {
		if(loadScene)
			throw new IllegalArgumentException("load scene on load scene");
		loadScene = true;
		
		clearScene();
		
		currentWorld = new World();
		currentWorld.add(new EnginePropertiesWorldComponent(properties));
		
		for(var property : properties)
			if(property instanceof WorldComponent c)
				currentWorld.add(c);
			
		currentSystems = nextScene.getSystems(globalSystems);
		if(currentSystems instanceof InitSystem isystem)
			isystem.init(currentWorld);
		
		nextScene.build(currentWorld);
		
		nextScene = null;
		
		loadScene = false;
	}
	
}

File: MoonEngine-Root/Base/src/main/java/com/greentree/engine/moon/base/scene/base/SceneManagerBase.kt
package com.greentree.engine.moon.base.scene.base

import com.greentree.commons.util.time.PointTimer
import com.greentree.engine.moon.base.scene.SceneManager
import com.greentree.engine.moon.ecs.CollectionWorld
import com.greentree.engine.moon.ecs.scene.MapSceneProperties
import com.greentree.engine.moon.ecs.scene.ProxySceneProperties
import com.greentree.engine.moon.ecs.scene.Scene
import com.greentree.engine.moon.ecs.scene.WorldProperty
import com.greentree.engine.moon.ecs.system.ECSSystem
import com.greentree.engine.moon.ecs.system.FullSystem
import com.greentree.engine.moon.ecs.system.toFull
import com.greentree.engine.moon.modules.property.EngineProperties
import org.apache.logging.log4j.LogManager
import java.util.*

class SceneManagerBase(private val properties: EngineProperties) : SceneManager {

	companion object {

		private val LOG = LogManager.getLogger(SceneManagerBase::class.java)
	}

	private val globalSystems: MutableCollection<ECSSystem> = ArrayList()
	private var nextScene: Scene? = null
	private var currentSystems: FullSystem? = null

	init {
		val autoAddGlobalSystems = ServiceLoader.load(ECSSystem::class.java)
		for(system in autoAddGlobalSystems) globalSystems.add(system)
	}

	@Synchronized
	override fun set(scene: Scene) {
		nextScene = scene
	}

	override fun addGlobalSystem(system: ECSSystem) {
		globalSystems.add(system)
	}

	fun update() {
		if(nextScene != null) loadScene()
		currentSystems!!.update()
	}

	@Synchronized
	private fun loadScene() {
		clearScene()
		val sceneProperties = ProxySceneProperties(properties, MapSceneProperties())
		val currentWorld = CollectionWorld()
		sceneProperties.add(WorldProperty(currentWorld))
		run {
			val timer = PointTimer()
			timer.point()
			currentSystems = nextScene!!.getSystem(globalSystems).toFull()
			timer.point()
			LOG.info("scene load ${timer[0]}")
		}
		nextScene = null
		run {
			val timer = PointTimer()
			timer.point()
			currentSystems!!.init(sceneProperties)
			timer.point()
			LOG.info("scene init ${timer[0]}")
		}
	}

	fun clearScene() {
		currentSystems?.destroy()
	}
}